unix/tclUnixThrd.c

```diff
@@ -159,11 +159,11 @@ Tcl_JoinThread(
 {
 #ifdef TCL_THREADS
     int result;
-    unsigned long retcode, *retcodePtr = &retcode;
+    void *retcode, **retcodePtr = &retcode;
```

Comment:
I think using `uintptr_t` for `retcode` and `void **` for `retcodePtr` would be cleaner here. Semantically, we want a variable that holds an integer that is of the same size as a pointer, i.e. we don't expect it to hold a valid capability but the pthread API requires a variable to be of that size. In result, when statically analysing this code, storing an invalid capability to this variable shouldn't be considered as an issue. Unfortunately, `uintmax_t` isn't suitable here for other reasons and I'm not aware of any data type that would be better here, i.e. an integer data type of the pointer size that cannot carry a valid pointer.

Reply:
I agree that, in general, `intptr_t` would be the right type for the task, but after I studied the project a bit, I decided to go with `void*` for the following reasons.

Return value is originally written by `TclpThreadExit` (next function in this file),
```c
pthread_exit(INT2PTR(status));
```
so this value will be converted through `void*` anyway due to pthread API.
```c
void pthread_exit(void *retval);
```

Furthermore, `INT2PTR` is defined at [tclInt.h:108](https://github.com/CTSRD-CHERI/tcl/blob/4d8b9b5b2f80db802357833a4d66062386889969/generic/tclInt.h#L108C1-L125C7),

```c
#if !defined(INT2PTR) && !defined(PTR2INT)
#   if defined(HAVE_INTPTR_T) || defined(intptr_t)
#	define INT2PTR(p) ((void *)(intptr_t)(p))
#	define PTR2INT(p) ((intptr_t)(p))
#   else
#	define INT2PTR(p) ((void *)(p))
#	define PTR2INT(p) ((int)(p))
#   endif
#endif
```

Implementation of these macros suggests that we must not rely on the availability of `intptr_t`, and will need to consider other cases, which will result in cluttering the code.

Regarding static analysis, it does not complain on integer to `void*` conversion in this function, but picks up an alignment issue (`unsigned long*` may not be sufficiently aligned to hold `void*`). 
As a side note, I think `pthread_exit(INT2PTR(status));` deserves an invalid capability warning (which cannot be fixed), even if there is no intention to dereference this value.

Anyway, I'm open to discussion :slightly_smiling_face:

Reply:
```int pthread_join(pthread_t, void **);``` is the prototype for the API in question here. Yes, the thread function is really returning an integer (or at least we hope so), so it's not a pointer, but that's not the point. The API is for `void **`, which means we need to pass a pointer to a `void *` to pthread_join, and casting `uintptr_t *` to `void **` is only marginally better than `unsigned long *` (in that it's at least the same underlying representation for CHERI), but it's still UB in ISO C as it violates the standard's aliasing requirements. So the caller here must pass in a "true" `void **`, after which point it is in control of the types and can massage the variable into whatever form it wants, in this case an int.

```pthread_exit(INT2PTR(status));``` shouldn't warn, INT2PTR in TCL is for "put an integer in something that needs to be of pointer type but isn't expected to be dereferenceable", and this is exactly that kind of case. That's why it's cast via intptr_t, as the idiomatic way to flag that you're putting an integer in a pointer deliberately.

Reply:
> `int pthread_join(pthread_t, void **);` is the prototype for the API in question here. Yes, the thread function is really returning an integer (or at least we hope so), so it's not a pointer, but that's not the point. The API is for `void **`, which means we need to pass a pointer to a `void *` to pthread_join, and casting `uintptr_t *` to `void **` is only marginally better than `unsigned long *` (in that it's at least the same underlying representation for CHERI), but it's still UB in ISO C as it violates the standard's aliasing requirements. So the caller here must pass in a "true" `void **`, after which point it is in control of the types and can massage the variable into whatever form it wants, in this case an int.

My thinking was somehow similar: the callers and callees of `pthread_join()` can use their own data types for a variable that becomes the `void **value_ptr` argument but in the end a valid `void **` argument must be constructed for the sake of the `pthread_join()` call. However, it is unclear if you suggest here to use `void *` for `retcode` or only make sure `retcodePtr` is `void **`. For me `uintptr_t` seemed more appropriate for `retcode` as it's controlled by the caller but I wanted `retcodePtr` to be `void **` to comply with the API.

On the other hand, Irina has pointed out that `intptr_t` might not be available, which is already handled with an ifdef within `INT2PTR()`. It would be terrible to use an ifdef in `Tcl_JoinThread()` that is in question here.

Reply:
I have talked with Irina and I understand that the intention here is to use the same data type for the `value_ptr` argument that `pthread_join()` operates on internally, which is `void *`, even though the caller of `pthread_exit()` understands it must cast the result to the integer data type. The storage of the `value_ptr` variable should be the same as seem by the pthread library and the user of pthread can allocate a variable of its own variable before it's stored in a variable to which a pointer is passed to pthread.